docs/objective-c_abi.md

```diff
@@ -0,0 +1,348 @@
+    Objective-C 2.0 features. Used on macOS 32-bit
+  * Modern ABI, version 2 - The modern 64-bit ABI. Used on all other Apple
+    platforms
+* GNU runtime - used on non-Apple platforms
+
+This document describes the Apple runtime with the modern ABI on macOS x86-64,
+as implemented by the Apple LLVM compiler (`9.0.0 (clang-900.0.39.2)`) shipped
+with Xcode 9.2. The information in this document has been obtained by reading
+documentation provided by Apple, looking at assembly outputs and object dumps
+from the LLVM compiler.
+
+Objective-C is a superset of C, therefore any of the language constructs that
+also exists in C, like functions, structs and variables use the C ABI of the
+platform. This document only describes the ABI of the Objective-C specific
+language constructs.
+
+## Messages
+
+The Objective-C model of object-oriented programming is based on message passing
+to object instances. Unlike D or C++ where a method is called. The difference
+from implementation stand point is that in D and C++ a vtable is used which
+is an array of function pointers and the compiler uses an index into that array
+to determine which method to call. In Objective-C it's the runtime that is
+responsible to find the correct implementation when a message is sent to an
+object. A method is identified by a *selector*, a null terminated string
+representing it's name, which maps to a C function pointer that implements the
+method.
+
+## Message Expression
+
+In Objective-C, sending a message to an object looks like the following example:
+
+```objective-c
+int result = [receiver message];
+```
+
+In D it would look like:
+
+```d
+int result = receiver.message();
+```
+
+The compiler implements this by making a regular C call to the `objc_msgSend`
+function in the Objective-C runtime. The signature of `objc_msgSend` looks
+something like this:
+
+```objective-c
+id objc_msgSend(id self, SEL op, ...);
+```
+
+* The first parameter is the receiver (this/self pointer)
+* The second parameter is the name of the method mentioned in the message -
+  that is, the method selector
+* The last parameter is for all the arguments that the
+  implementation expects
+
+The above example would be translated by the compiler to the following:
+
+```c
+int result = objc_msgSend(receiver, selector);
+```
+
+If the method expects any arguments they're passed after the selector argument:
+
+```c
+int result = objc_msgSend(receiver, selector, arg1, arg2);
+```
+
+The call to `objc_msgSend` should not be performed as a variadic call but
+instead as if it had the same signature as the method that should be called but
+with the two additional parameter, `self` and `op`, added first. The
+implementation of `objc_msgSend` will jump to the method instead of calling it.
+
+Because of the above, multiple versions of `objc_msgSend` exist. Depending on
+the return type of the method that is called the correct version need to be
```

Comment:
comma after "called". 

s/the correct version need to be used/the correct version will be used